### PR TITLE
Remove outdated docs for training_step

### PR DESCRIPTION
**Pull request by quinor**
## What does this PR do?

Skipping batch by returning `None` from `training_step` wasn't working properly in some cases, ie. when using `torch.cuda.amp`. Since the introduction of `LightningOptimizer`, I believe the supported/recommended method is by using manual optimization (docs still pending). I'm proposing to remove the docs from `training_step` (still keep the functionality for backwards compatibility though).

I think removing it is the right move, but an argument can be made to make it deprecated instead. I don't like it since the feature does not always work anyway.

Fixes # (issue)

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Comment by carmocca** [comment]
> I think removing it is the right move, but an argument can be made to make it deprecated instead. I don't like it since the feature does not always work anyway.

If it is not working properly then we should try to fix it or otherwise warn the users about invalid configurations, but there are users who rely on this feature and whose particular use-case does work.

> wasn't working properly in some cases

Can you expand on these cases? I would prefer if you opened a bug report (ideally with a reproducible example) for these instead of just removing it from the docs.

**Comment by quinor** [comment]
@carmocca The case that touched me personally was it not working when using native 16-bit precision on a GPU, that is pytorch AMP. I did in fact open a bug report (https://github.com/PyTorchLightning/pytorch-lightning/issues/4524) and the resolution was "won't fix, manual optimization is recommended in such case". Since 16-bit precision is advertised as working out-of-the-box, my argument is that it's very bad for one easy feature to break another easy feature which is used for a wide array of models such as GANs.

I'm not 100% sure, but I've heard that similar interactions may also happen with other accelerations that have some "special needs" and I think the general resolution is to use LightningOptimizer to wrap all of those special needs (@tchaton is working on that right now).

As for the reproduction, please see https://colab.research.google.com/drive/1Fi-r5twOjag0YKrIy2SxqMAAMf3zX5xj#scrollTo=Flyi--SpvsJN

**Comment by carmocca** [comment]
Sorry, I did not have the context of the issues you linked. However, let's not mix different issues:

1. Issues with AMP and overriden `optimizer_step` (#4572). Seems to have been addressed in that issue
2. Issues with AMP and skipping `training_step` (#4524). I agree here that it has not been resolved and should be reopened.

Now, about this docs update, we could add a note about AMP not supporting skipping a step (not sure if it can be fixed, we can continue that discussion in #4524). But it doesn't make sense to remove it entirely.

**Comment by quinor** [comment]
Yeah, 1) (#4572) was fixed, though both bugs were just issues I encountered trying to implement the same behavior.

Hmm, the main reason I wanted to remove it is that I was under the impression it's not recommended to do it that way and that manual optimization is the recommended way instead. If you manage to fix #4524 (and make it robust for other cases as well!) that would be a great solution. I tried doing that but failed to find a simple solution, most of the things I know are in the discussion of #4524.

**Review by SeanNaren — APPROVED** [review]

**Comment by carmocca** [comment]
@Borda @SeanNaren I dont think we want to merge this for the reasons discussed in this PR

**Comment by SeanNaren** [comment]
Went down the trail of issues, thanks for the links @carmocca!

Returning none from the training step has lead to tricky issues hence why I was fine with removing it. But on the flip side it should just work, as users expect it to (skip the training step without having to define manual optimization).

I think with the latest LightningOptimizer changes the original issue #4524 could be easier to address. Do you mind having a look @carmocca?

Just to be clear the logic should be: training step returns none -> skip optimization step?

**Comment by carmocca** [comment]
> Went down the trail of issues, thanks for the links @carmocca!

Thanks for taking the time!

> I think with the latest LightningOptimizer changes the original issue #4524 could be easier to address. Do you mind having a look @carmocca?

Yes, i'll try since this issue matters to me. (my research requires skipping training steps).

> Just to be clear the logic should be: training step returns none -> skip optimization step?

Yes. Usually one returns None due to an invalid loss value (NaN, inf...) so there is nothing to run backward on.
